### PR TITLE
Issue 3338: lmsig_ksils12: Add high avalability mode supported by libksi 3.18

### PR DESCRIPTION
**Pull request by allanpark**
Libksi version 3.18 supports high availability mode of operation. In HA mode multiple server endpoints can be configured and requests are sent to all them simultaneously while the first valid response is always used. This improves the availability of the service and reduces response latency.

### CODE REVIEW

**Comment by rgerhards** [comment]
@allanpark I assume we need newer libksi on the failing machines? If so, I think the libksi version check needs a bump - we should not go past configure if we know that the compile will fail with an older version.

**Comment by allanpark** [comment]
I fixed the libksi vesrsion in configure.ac to 3.18 but this wasn't the issue for docker anyway. The docker scripts in rsyslog-docker repository use git clone && make and everything should be ok. But git is somehow checking out the previous version from github although 3.18 is in the master branch. 

Are there any other logs from docker build available?

**Comment by rgerhards** [comment]
I should add: we also do not WANT to always build the latest version from git. If we did so, we would not be able to detect bugs in the version requirement in `configure.ac` (and distro maintainers would complain later on). So it is not just an operational issue that the containers are not re-created each time.

**Comment by rgerhards** [comment]
@allanpark while we are at it: is there any way to add automatted tests for the KSI component? I originally raised the question in https://github.com/rsyslog/rsyslog/issues/3073 and would like to nail it (if the answer is "no", then let's log this and exclude it from coverage with good feeling).

**Comment by rgerhards** [comment]
@allanpark while you are at it, it would be great if you could squash the commits. As far as I can see, they are all fixes to prior commits in the same set. I can also squash as part of the merge, but you obviously have more control over commit text etc if you do it. So no hard requirement that you actually do it.

**Comment by allanpark** [comment]
@rgerhards The file was actually duplicated by accident and I shall fix it in the branch. Squashing everything also makes sense since it is all actually about one feature. I shall fix this. 

Actually I have added a test script some time ago (tests/lmsig_ksi_ls12.sh, tests/resultdata/lmsig_ksi_ls12_async/*). For testing high availability mode a new test should be added (or the current one updated), I can do this in a separate issue/patch.

**Comment by allanpark** [comment]
I shall close this pull request and create another with all the commits squashed into one for clarity.

**Comment by rgerhards** [comment]
@allanpark to not loose history, I would prefer it you could push the squased version into the same branch/PR. Not a hard requirement, but much better from a management PoV. :-)

**Comment by rgerhards** [comment]
@allanpark all you actually need to do is

* squash (git rebase -i HEAD~nn)
* git push -f 

The rest happens automatically if the PR is reopened and the branch remains the same.

**Comment by allanpark** [comment]
Alright, I shall do that!

**Comment by rgerhards** [comment]
@allanpark just a friendly reminder: merge window for 2019.01 release closes next week.

**Comment by allanpark** [comment]
> @allanpark just a friendly reminder: merge window for 2019.01 release closes next week.
@rgerhards I have fixed the issue (removed the accidentally added file) and squashed the commits. - but please hold the the merge back for time being because we seem to have found a last minute issue with the patch. If this delay will cause us to miss the merge window this time then let's better wait for the next release.

**Comment by rgerhards** [comment]
> but please hold the the merge back for time being because we seem to have found a last minute issue with the patch. If this delay will cause us to miss the merge window this time then let's better wait for the next release.

thx for letting me know. Please tell me explicitly when it is OK to merge (and feel free to push interim updates to run CI).

**Comment by rgerhards** [comment]
@allanpark just so that you know: I am *probably* able to do late-merges on monday. Again, just mentioning, don't feel pressed to anything (and I cannot guarantee Monday!).

**Comment by rgerhards** [comment]
@allanpark friendly reminder: merge window for 8.1903.0 closes Friday.

**Comment by allanpark** [comment]
> @allanpark friendly reminder: merge window for 8.1903.0 closes Friday.

@rgerhards Thanks for the reminder. We have solved the final issues and are ready to merge now. But there seems to be an issue with a dependency again: libksi 3.19 is not found. Could it be again that the old version is cached?

**Comment by rgerhards** [comment]
@allanpark 

> Could it be again that the old version is cached?

There is a misunderstanding on your part: libksi is a **dependency component**. Those are always installed into the CI env and need to be requested to be upgraded. Nothing of that happens, and does not do so intentionally.

**Comment by rgerhards** [comment]
That said... the containers were recently rebuild. When was the new version released?

**Comment by allanpark** [comment]
> That said... the containers were recently rebuild. When was the new version released?

Sorry, I missed the container part, but the all the other buildbots seem to be ok. The github code was apudated 23 day ago and the packages were updated even before that. But how does it look, is it still possible to do the merge this time?

**Comment by rgerhards** [comment]
Well not all machines execute the same set of tests. I think KSI is disabled in most of the CI runs, because we unfortunately have no dynamic tests, so there is no point in wasting time on building something that can not be dynamically checked in any case. Each CI env is unique, and for a "go", basically all need to pass. You often see different issues in different envs.

I'll try to re-build the container, but I was sure I did this not longer than 2 weeks ago. Let's see, I keep you posted.

File: configure.ac [hunk 1]
```diff
@@ -1,4 +1,5 @@
-#                                               -*- Autoconf -*-
+i
```
Comment (rgerhards) [inline configure.ac]: this doesn't look right ;-)

Comment (allanpark) [reply configure.ac]: Fixed!

**Comment by rgerhards** [comment]
@allanpark just to let you know: container build turns out to be more work than anticipated due to some dependencies. But I am not sure I manage to complete it today or early tomorrow. The warning phase was a bit short ;-)

**Comment by rgerhards** [comment]
As a side-note, I usually need 1 to 2 weeks advance notice for component bumps.

**Comment by allanpark** [comment]
@rgerhards Just humbly asking: are there still chances to get this merged before the release? 
And of course - apologies for the last minute hassle... :-)

**Comment by rgerhards** [comment]
@allanpark I try, probably over the weekend. I have one issue with imfile that really needs a fix with the next release and for now I need to concentrate on that. For the containers: if all goes well, it should be no issue, but not always all goes well. And as you know, the more you need it, the less it goes well. Usually I can do the updates within a day. But every now and then updated components cause trouble, and this can be real grief. Just had it recently when we did the last update round. That's why I need to have 1 to 2 weeks advance notice. But now let's see how things evolve over the weekend...

**Comment by rgerhards** [comment]
looks like the container goods were with me. Now doing close/re-open trying to give CodeFactor a new push ;-)

**Pull request merged**